### ISSUE TITLE
unittest.wait() calls runUntilCurrent recursively

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10508]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#781|
|Type|defect|
|Created|2004-11-08 01:25:21Z|

<details><summary>Searchable metadata</summary>

```
trac-id__781 781
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1099877121000000 1099877121000000
changetime__1128341348000000 1128341348000000
version__ 
owner__ 
cc__exarkun cc__jml cc__slyphon
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
This causes problems for tests that use threads, eg test_adbapi.  I've added two
test cases that demonstrate the problem to test_trial.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
it would be a big help to me if you included a docstring or comment explaining
what these tests are doing to cause the situation to occur
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
I don't believe i can make wait() reentrant without using threads, so:

wait keeps a counter to make sure that it is only waiting on one deferred
at a time and is not reentered. If it is reentered, it raises a 
util.WaitIsNotReentrantError. there are no tests in twisted that currently 
rely upon recursive calls to wait, so I am opting to say "Don't do that" 
for the moment, and coming up with the nasty "real" answer if a valid 
use-case surfaces.

committed in revision 12402
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
oh, and the unittests that exarkun so kindly added have been incorporated and
now test to make sure that the methods he wrote raise util.WaitIsNotReentrantError
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Please update the trial documentation to include something along these lines:

  When returning a Deferred from a setUp, tearDown, or test method to trial, it
is important to note that there may be some restrictions placed on the callbacks
which may appear on this Deferred.  Specifically, depending on the
implementation of API which created them, they may not be allowed to invoke
deferredResult, deferredError, or wait.  Whether they can or cannot is
unknowable from only the interface and, as mentioned above, depends solely on
the implementation which is responsible for firing the Deferred.  If the methods
are restricted, a WaitIsNotReentrantError exception will be called upon their
invocation.

  Since even when the methods do work, changes to their implementation which do
not otherwise effect their interface could break unit tests which rely on
deferredResult, deferredError, and wait, it is recommended that none of these
methods be used in _any_ callback on a Deferred which is returned to trial.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="slyphon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|slyphon commented|
|-|-|

```
#!html
<pre>
requested documentation update committed 12405

updated both wait() docstring and unittest module docstring
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
what's this issue blocking on now? it looks like it's resolved
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
New maintainer.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
Everything seems to be in order.  Marking as resolved
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
